lib/src/annotate.rs

```diff
@@ -0,0 +1,335 @@
+        results
+    }
+
+    fn forward_to_new_commit(
+        &mut self,
+        old_commit_id: &CommitId,
+        old_local_line_number: usize,
+        new_commit_id: &CommitId,
+        new_local_line_number: usize,
+    ) {
+        let old_map = self.local_line_map.get_mut(old_commit_id).unwrap();
+        let removed_original_line_number = old_map.remove(&old_local_line_number);
+        if removed_original_line_number.is_none() {
+            return;
+        }
+        let removed_original_line_number = removed_original_line_number.unwrap();
+        if self.local_line_map.contains_key(new_commit_id) {
+            self.local_line_map
+                .get_mut(new_commit_id)
+                .unwrap()
+                .insert(new_local_line_number, removed_original_line_number);
+        } else {
+            let mut new_map = HashMap::new();
+            new_map.insert(new_local_line_number, removed_original_line_number);
+            self.local_line_map.insert(new_commit_id.clone(), new_map);
+        }
+    }
+
+    fn swap_full_commit_id(&mut self, old_commit_id: &CommitId, new_commit_id: &CommitId) {
+        let old_commit_map = self.local_line_map.remove(old_commit_id);
+        if old_commit_map.is_none() {
+            return;
+        }
+        let old_commit_map = old_commit_map.unwrap();
+        self.local_line_map
+            .insert(new_commit_id.clone(), old_commit_map);
+    }
+
+    fn drain_remaining_for_commit_id(&mut self, commit_id: &CommitId) {
+        let remaining_lines = self.local_line_map.remove(commit_id);
+        if remaining_lines.is_none() {
+            return;
+        }
+        let remaining_lines = remaining_lines.unwrap();
+        for (_, original_line_number) in remaining_lines {
+            self.original_line_map
+                .insert(original_line_number, commit_id.clone());
+        }
+    }
+
+    fn get_next_commit(&self) -> Option<CommitId> {
+        return self.local_line_map.keys().next().cloned();
+    }
+
+    fn convert_to_results(self, original_contents: &[u8]) -> AnnotateResults {
+        let original_content_lines: Vec<String> =
+            original_contents.lines().map(|s| s.unwrap()).collect();
+        let mut result_lines = Vec::new();
+        for (idx, line) in original_content_lines.into_iter().enumerate() {
+            result_lines.push((self.original_line_map.get(&idx).unwrap().clone(), line));
+        }
+        AnnotateResults {
+            file_annotations: result_lines,
+        }
+    }
+}
+
+/// Get line by line annotations for a specific file path in the repo.
+pub fn get_annotation_for_file(
+    file_name: &str,
+    repo: &ReadonlyRepo,
+    starting_commit_id: &CommitId,
+) -> Result<AnnotateResults, AnnotateError> {
+    let file_path = RepoPathBuf::from_relative_path(file_name);
+    if file_path.is_err() {
+        return Err(AnnotateError::FileNotFound(file_name.to_string()));
+    }
+    let file_path: RepoPathBuf = file_path.unwrap();
```

Comment:
nit: for type safety, `file_name` can be converted to `RepoPath` by caller.

Reply:
done